Create a table component

fhlavac opened the pull request:
[HAC-753](https://issues.redhat.com/browse/HAC-753)

Table:
![table1](https://user-images.githubusercontent.com/50696716/156834009-44e9c45e-c405-489e-a09a-60ed57f3344a.png)

Loading:
![loading](https://user-images.githubusercontent.com/50696716/156834031-ab3ad810-05e1-42dd-9243-c449f16f8214.png)

Default error:
![error](https://user-images.githubusercontent.com/50696716/156834056-7e7f6dfa-1276-4485-8610-7bcc85a07258.png)

Data:
```
const columns = [
    {
      title: 'Name',
      id: 'name',
      sort: 'metadata.name',
      transforms: [sortable], 
      props: {
        className: '',
      },
    },
    {
      title: 'Namespace',
      id: 'namespace',
      props: {
        className: '',
      },
    },
    {
      title: 'Status',
      id: 'status',
      transforms: [null],
      props: {
        className: '',
      },
    },
    {
      title: 'Detail',
      id: 'detail',
      transforms: [sortable],
      props: {
        className: '',
      },
    },
    {
      title: 'Note',
      id: 'note',
      transforms: [null],
      props: {
        className: '',
      },
    },
  ];

const data = [
  { name: 'one', namespace: 'two', status: 'three', detail: 'four', note: 'five' },
  { name: 'one - 2', namespace: null, status: null, detail: 'four - 2', note: 'five - 2' },
  {
    name: 'one - 3',
    namespace: 'two - 3',
    status: 'three - 3',
    detail: 'four - 3',
    note: 'five - 3',
  },
];

export const Row: React.FC<RowProps<RowObject, Record<string, unknown>>> = ({ obj }) => (
  <>
    <Td dataLabel={obj.name}>
      {obj.name}
    </Td>
    <Td dataLabel={obj.namespace}>
      {obj.namespace}
    </Td>
    <Td dataLabel={obj.status}>
      {obj.status}
    </Td>
    <Td dataLabel={obj.detail}>
      {obj.detail}
    </Td>
    <Td dataLabel={obj.note}>
      {obj.note}
    </Td>
  </>
);
```

Component usage:
```
<SdkTable
  columns={columns}
  data={data}
  loaded={false}
  // emptyLabel={'Something went wrong...'}
  // loadError={'An error occured when fetching the data'}
  Row={Row}
/>
```

florkbr commented on packages/lib-utils/src/components/status/StatusBox.tsx:
```diff
@@ -0,0 +1,72 @@
+// @ts-nocheck
```
What issues did we run into with typescript here? I don't think we could merge this without TS checking.

fhlavac replied:
This code is so far just a draft. I'll remove this. It's just for dev purposes to prevent it from causing build errors.

florkbr commented on packages/lib-utils/src/components/status/StatusBox.tsx:
```diff
@@ -0,0 +1,72 @@
+import * as _ from 'lodash';
+
+type AccessDeniedProps = {
+  message?: string;
```
Wondering if we should have a fallback message if none are provided? Or we could hide the error details section if no message is provided. I'm thinking it would look odd to see `Error details` in the UI and underneath is blank :)

fhlavac replied:
Absolutely, I'll add it.

florkbr commented on packages/lib-utils/src/components/status/StatusBox.tsx:
```diff
@@ -0,0 +1,72 @@
+
+  if (loadError) {
+    const status = _.get(loadError, 'response.status');
+    if (status === 404) {
```
we could probably do a `switch` statement on the `status` so it's easier to add/catch new cases in the future.

florkbr commented on packages/lib-utils/src/components/table/Table.tsx:
```diff
@@ -0,0 +1,95 @@
+// @ts-nocheck
```
Same here, should re-enable TS when we can.

florkbr commented on packages/lib-utils/src/components/table/Table.tsx:
```diff
@@ -0,0 +1,95 @@
+  Th,
+  Thead,
+  Tr,
+} from '@patternfly/react-table';
```
Did we try starting with https://www.patternfly.org/v4/components/table/ first and tweak it to what we need? I'm worried we're rebuilding another table implementation using the patternfly building blocks. That might be necessary in the end, but we should start simple.

 I think the idea is we want a `Table` component as a middle man between SDK consumers and the Patternfly table that can sprinkle some K8s handling magic inbetween.

fhlavac replied:
Not sure I'm getting your point. Do you mean using `<Table>` PF component instead of `<TableComposable>` with the build blocks? I wanted to avoid using `<Table>` since it's marked as legacy - see https://www.patternfly.org/v4/components/table/react-legacy

florkbr replied:
Ah, I completely missed that! Good point. I didn't see the `TableComposable` import

fhlavac commented:
@florkbr thanks for the review. Sorry for the small errors and TS ignores. I should have warned you :laughing: 
I wanted more of a review of the concept and features we need urgently than code details. I'll go through it a lot of times again and clean it. Just wanted to be clear about what is a priority for us and if this is ok from the concept point of view.

florkbr commented on packages/lib-utils/src/components/table/Table.tsx:
```diff
@@ -0,0 +1,95 @@
+  data: D[];
+  unfilteredData: D[];
+  loaded: boolean;
+  loadError: any;
```
Since we have a chance to revisit our table component API, we should try avoiding `any` if possible

fhlavac commented:
Some more questions:


**1) Ad StatusBox** - status component wrapping the table
- I added just very simple statuses, that comply with the virtualized table status types. How much do we want to bring from the console's version in this implementation from the UX point of view? 
- Do we want to use some kind of table skeleton or the empty state is currently enough?
- `IncompleteDataError`, `TimeoutError` do we want to migrate them from console/dynamic-plugin-sdk? For now commented the code sections that should use them.


**2) Ad types**

@vojtechszocs disabled `react/jsx-props-no-spreading`, `'react/require-default-props`. Is it ok or do we want to have any of these enabled?


**3) Ad table**
- do we want to use `rowData` attribute?
- do we want to use `activeColumnIds` attribute now?
- not sure what `additional` is in `TableColumn` type
- for table props - do we want to deal with `scrollNode`, `gridBreakPoint`
- do we want to utilize `onSelect` - selecting items?

fhlavac commented:
@florkbr pushed updates according to our sync on Tuesday. Cleaned the code, modified component properties, and added a possibility to use custom sorting.

**In follow up PRs, I think we should:**
- add filtering
- add row selecting
- handle translations
- use rowRenderer - `rowData` property
- manage columns being displayed
- handle scrolling and pagination

florkbr commented on packages/eslint-plugin-internal/src/rules/react-typescript-prettier.js:
```diff
@@ -10,4 +10,10 @@ module.exports = {
+  'react/jsx-props-no-spreading': 'off',
+
+  // Disable default props require
+  'react/require-default-props': 'off',
```
@fhlavac I really think we should keep these rules enabled if possible. 

`jsx-props-no-spreading` - while I like this for easy component development, we need to think of our components as an API, and tend to agree with the eslint doc:
> This enhances readability of code by being more explicit about what props are received by the component. It is also good for maintainability by avoiding passing unintentional extra props and allowing react to emit warnings when invalid HTML props are passed to HTML elements

`require-default-props` - I think this forces us to consider which props are required and which are not WRT our component API.

> One advantage of defaultProps over custom default logic in your code is that defaultProps are resolved by React before the PropTypes typechecking happens, so typechecking will also apply to your defaultProps.

fhlavac replied:
Completely agree with the `require-default-props` just not sure about `jsx-props-no-spreading ` since in https://github.com/openshift/dynamic-plugin-sdk/pull/34/files#diff-b56b27c61ebe92f1fe8626d77d6dcae22ddfbccfd4c1d71f7da9d3c629e615c5R101 we probably want to spread any properties passed. Disabled the rule for that line for now.

florkbr commented on packages/lib-utils/package.json:
```diff
@@ -24,6 +24,8 @@
   "peerDependencies": {
     "@openshift/dynamic-plugin-sdk": "^1.0.0",
+    "@patternfly/react-core": "^4.198.19",
+    "@patternfly/react-table": "^4.67.19",
```
@fhlavac let's leave this as is for now. I'll need to think on this a bit more. Technically I believe we can have multiple different versions of a patternfly component running on a page - therefore we may not need to enforce the table component as a peer dependency.

florkbr commented on packages/lib-utils/src/components/table/Table.tsx:
```diff
@@ -0,0 +1,126 @@
+  sort?: ThProps['sort'] | ((data: D[], sortDirection: SortByDirection) => D[]) | string;
+};
+
+type TableProps<D> = {
```
@fhlavac we should also export the table props object to be used when creating an instance of the table.

florkbr commented on packages/lib-utils/src/components/status/StatusBox.tsx:
```diff
@@ -0,0 +1,99 @@
+  </EmptyState>
+);
+
+type StatusBoxProps = {
```
We should export the type object for status box props too

florkbr commented on packages/lib-utils/src/components/status/StatusBox.tsx:
```diff
@@ -0,0 +1,99 @@
+
+type StatusBoxProps = {
+  areFiltersApplied?: boolean;
+  data?: unknown;
```
@fhlavac FWIW `unknown` should be avoided. Can we use a shape?

fhlavac replied:
@florkbr I will probably use `Record<string, unknown>`. Is that ok?

andrewballantyne replied:
@fhlavac @florkbr I like to make a callback to my point around `unknown` I [made on slack](https://coreos.slack.com/archives/C02EQERSV3N/p1646162496757559). I still believe this is true... types need to be Generic long before we consider `unknown` valid. `unknown` is a step up from `any` and as an SDK we will definitely have `unknown`s... but we need to make sure this is not something that can be properly typed. Otherwise the code that receives `data` will be effectively `any`.

fhlavac replied:
@andrewballantyne @florkbr added a generic type to `StatusBoxProps`. In this case, the `data` property is used just to verify if it is empty or not so it's probably enough to expect that it is an array.

florkbr replied:
@andrewballantyne agreed; I pinned your comment in the forum as well so we can find it again :)

florkbr replied:
@fhlavac for now that's ok. One thing of note - if we're only checking `data` to see if it's empty, we could just add a new prop `empty: boolean;` and do the check in the consumer (that way we don't need to know anything about the data structure here).

fhlavac commented on packages/lib-utils/src/components/status/StatusBox.tsx:
```diff
@@ -0,0 +1,99 @@
+  message = 'Insufficient access rights.',
+}) => (
+  <EmptyState>
+    <Title headingLevel="h2">Restricted Access</Title>
```
@florkbr created one https://issues.redhat.com/browse/HAC-958

fhlavac commented:
/retest

vojtechszocs commented on packages/lib-utils/src/components/status/StatusBox.tsx:
```diff
@@ -0,0 +1,99 @@
+  <EmptyState>
+    <Title headingLevel="h2">Restricted Access</Title>
+    <EmptyStateBody>
+      <Text>You don&apos;t have access to this section due to cluster policy.</Text>
```
If it makes sense, I'd suggest a follow-up change where we use props instead of hard-coded strings like the one above.

fhlavac replied:
I'll add this to a follow-up list. We probably have to discuss it a bit. In the original console's implementation, there is a translated string being rendered. Since we don't have translations yet, I put just this string for now.

vojtechszocs replied:
@fhlavac @florkbr Well, there are two aspects to this, first one is localization (for which we currently don't have support in the SDK repo yet), while the second one is reusability (hard-coded strings, regardless if localized or not, make a component less reusable).

Ideally, for every component we move from Console to `lib-utils` in the SDK repo, we should re-evaluate its API and code. I agree that initially we want to have some components ready and good-enough to be used by HAC Core, this is more for a long-term component & API related discussion.

vojtechszocs commented on packages/lib-utils/src/components/status/StatusBox.tsx:
```diff
@@ -0,0 +1,99 @@
+  emptyLabel?: string;
+  loadError?: LoadError;
+  loaded?: boolean;
+  LoadErrorDefaultMsg?: React.ComponentType<Record<string, never>>;
```
`Record<string, never>` infers into an empty object, see [here](https://github.com/typescript-eslint/typescript-eslint/issues/2063#issuecomment-675156492) for details.

I'd suggest adding the following into `packages/common/src/types/common.ts` right after `AnyObject` type declaration:
```ts
/**
 * Use the `EmptyObject` type for objects that have no properties.
 * 
 * @see https://github.com/typescript-eslint/typescript-eslint/issues/2063#issuecomment-675156492
 */
export type EmptyObject = Record<string, never>;
```

and using `EmptyObject` wherever it's needed.

vojtechszocs commented on packages/lib-utils/src/components/status/StatusBox.tsx:
```diff
@@ -0,0 +1,99 @@
+  EmptyMsg?: React.ComponentType<Record<string, never>>;
+};
+
+export const StatusBox: React.FC<StatusBoxProps<Record<string, unknown>>> = ({
```
You can use the existing `AnyObject` type which is aliased to `Record<string, unknown>`.

```ts
import type { AnyObject } from '@monorepo/common';
```

I'd suggest doing a search & replace where
- `Record<string, unknown>` :arrow_right: `AnyObject`
- `Record<string, never>` :arrow_right: `EmptyObject`

fhlavac commented on packages/lib-utils/src/components/status/StatusBox.tsx:
```diff
@@ -0,0 +1,99 @@
+    );
+  }
+
+  if (!data || _.isEmpty(data)) {
```
then we return `children` - in this case the table. Is that right?

vojtechszocs replied:
Well, I was just curious as to why `data` is not used at all once we have it non-empty :smiley: just returning `children` is typically used as fallback behavior, normally we'd use `data` to render something different than `children` pass-through.

So if this is intentional and we want to have it working this way, we can keep it. I'd definitely put it on a follow-up TODO list to understand why this code is written this way, and maybe refactor to make it simpler.

fhlavac commented on packages/lib-utils/src/components/table/Table.tsx:
```diff
@@ -0,0 +1,122 @@
+  loadError?: LoadError;
+  columns: TableColumn[];
+  Row: React.FC<RowProps<D>>;
+  LoadErrorDefaultMsg?: React.ComponentType;
```
makes sense, thank you :slightly_smiling_face:

vojtechszocs commented:
@fhlavac One more change, please add

```json
"@patternfly/react-core": "^4.198.19",
"@patternfly/react-table": "^4.67.19",
```

also to `packages/lib-utils/package.json` into `peerDependencies` to ensure they are treated as proper external dependencies.

florkbr reviewed (approved):
Looks good! Thanks @fhlavac. We have more work to do on the table in the future, but this is a good start.

florkbr commented:
/lgtm

florkbr commented:
/approve
